### PR TITLE
refactor (Corners) change data class implementation

### PR DESCRIPTION
**Pull request by giovanni-nic-kuama**
address [ADS-14]

### CODE REVIEW

File: app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt [hunk 1]
```diff
@@ -68,7 +69,16 @@ class FindPaperSheetContours : UseCase<PaperSheetContoursResult, FindPaperSheetC
             }
 
             val result: PaperSheetContoursResult = contours.firstOrNull()?.let {
-                PaperSheetContoursResult(params.bitmap, Corners(it.shape.toList(), original.size()))
+                var corners = Corners(
```
Comment (matteoKuama) [inline app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt]: Maybe better to build corners already with sorted points?
`val corners = CornerFactory.createSorted(shape, size)`

File: app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt [hunk 2]
```diff
@@ -68,7 +69,16 @@ class FindPaperSheetContours : UseCase<PaperSheetContoursResult, FindPaperSheetC
...
+                    topLeft = it.shape[0],
+                    topRight = it.shape[1],
+                    bottomLeft = it.shape[2],
+                    bottomRight = it.shape[3],
+                    size = original.size()
+                )
+                corners = PerspectiveTransformUtils.sortPoints(corners)
```
Comment (matteoKuama) [inline app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt]:
```suggestion
                val corners = CornerFactory.createSorted(shape, size)
```

Comment (matteoKuama) [reply app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt]: or similar

**Review by matteoKuama — CHANGES_REQUESTED** [review]

File: app/src/main/java/net/kuama/documentscanner/data/Corners.kt [hunk 3]
```diff
@@ -1,12 +1,19 @@
...
-    var bottomLeft: Point = points[3]
+data class Corners(
+    val topLeft: Point,
+    val topRight: Point,
+    val bottomRight: Point,
+    val bottomLeft: Point,
+    val size: Size
+)
+
+class CornersFactory {
```
Comment (matteoKuama) [inline app/src/main/java/net/kuama/documentscanner/data/Corners.kt]:
```suggestion
object CornersFactory {
```

File: app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt [hunk 4]
```diff
@@ -68,7 +68,9 @@ class FindPaperSheetContours : UseCase<PaperSheetContoursResult, FindPaperSheetC
             }
 
             val result: PaperSheetContoursResult = contours.firstOrNull()?.let {
-                PaperSheetContoursResult(params.bitmap, Corners(it.shape.toList(), original.size()))
+                val corners = CornersFactory().create(it.shape, original.size())
```
Comment (matteoKuama) [inline app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt]:
```suggestion
                val corners = CornersFactory.create(it.shape, original.size())
```

File: app/src/main/java/net/kuama/documentscanner/utils/PointUtils.kt [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package net.kuama.documentscanner.utils
+
+import net.kuama.documentscanner.data.Corners
+import org.opencv.core.Point
+import org.opencv.core.Size
+import java.util.*
+import kotlin.Comparator
+
+object PointUtils {
+    fun getSortedCorners(it: Array<Point>, originalSize: Size): Corners {
```
Comment (matteoKuama) [inline app/src/main/java/net/kuama/documentscanner/utils/PointUtils.kt]:
```suggestion
    fun getSortedCorners(cornersList: Array<Point>, originalSize: Size): Corners {
```
usually `it` is the parameter inside of lambdas in kotlin

File: app/src/main/java/net/kuama/documentscanner/utils/PointUtils.kt [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        val cornerList = it.toList()
+
```
Comment (matteoKuama) [inline app/src/main/java/net/kuama/documentscanner/utils/PointUtils.kt]: check that the cornerList size is 4

**Review by matteoKuama — CHANGES_REQUESTED** [review]

File: app/src/main/java/net/kuama/documentscanner/utils/PointUtils.kt [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import net.kuama.documentscanner.data.Corners
+import org.opencv.core.Point
+import org.opencv.core.Size
+import java.util.*
+import kotlin.Comparator
+
+object PointUtils {
+    fun getSortedCorners(it: Array<Point>, originalSize: Size): Corners {
+        val cornerList = it.toList()
```
Comment (GabM3) [inline app/src/main/java/net/kuama/documentscanner/utils/PointUtils.kt]: if you convert it to a list immediatly, why don't you accept `List<Point>` in the constructor directly? isn't it better? 😄

**Review by GabM3 — CHANGES_REQUESTED** [review]

File: app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt [hunk 8]
```diff
@@ -61,15 +59,12 @@ class FindPaperSheetContours : UseCase<PaperSheetContoursResult, FindPaperSheetC
...
-            } else {
-                params.bitmap.recycle()
-            }
+            params.bitmap.recycle()
 
-            val result: PaperSheetContoursResult = contours.firstOrNull()?.let {
-                PaperSheetContoursResult(params.bitmap, Corners(it.shape.toList(), original.size()))
-            } ?: PaperSheetContoursResult(params.bitmap, null)
+            val result: Corners? = contours.firstOrNull()?.let {
+                val corners = CornersFactory.create(it.shape, original.size())
```
Comment (matteoKuama) [inline app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt]:
```suggestion
                CornersFactory.create(it.shape, original.size())
```

File: app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt [hunk 9]
```diff
@@ -61,15 +59,12 @@ class FindPaperSheetContours : UseCase<PaperSheetContoursResult, FindPaperSheetC
...
+                corners
```
Comment (matteoKuama) [inline app/src/main/java/net/kuama/documentscanner/domain/FindPaperSheetContours.kt]:
```suggestion

```

File: app/src/main/java/net/kuama/documentscanner/viewmodels/CropperModel.kt [hunk 10]
```diff
@@ -32,9 +30,8 @@ class CropperModel : ViewModel() {
                 )
             ) { either ->
                 either.fold(::handleFailure) { preview ->
-                    analyze(preview, returnOriginalMat = true) { pair ->
```
Comment (matteoKuama) [inline app/src/main/java/net/kuama/documentscanner/viewmodels/CropperModel.kt:35]: without value, what is the default? here it seems true, in the signature of analyze false..
it is important to decide what is the correct behavior now in the usecase

**Review by matteoKuama — CHANGES_REQUESTED** [review]

**Review by matteoKuama — APPROVED** [review]

**Review by GabM3 — APPROVED** [review]

**Pull request merged**